R/busy-indicators.R

```diff
@@ -117,49 +128,86 @@ useBusyIndicators <- function(..., spinners = TRUE, pulse = TRUE) {
 #' shinyApp(ui, server)
 busyIndicatorOptions <- function(
   ...,
+  spinner_type = NULL,
   spinner_color = NULL,
   spinner_size = NULL,
   spinner_delay = NULL,
-  spinner_selector = NULL,
+  spinner_selector = ":root",
   pulse_background = NULL,
   pulse_height = NULL,
   pulse_speed = NULL
 ) {
 
   rlang::check_dots_empty()
 
-  res <- tagList(
-    spinnerOptions(
-      color = spinner_color,
-      size = spinner_size,
-      delay = spinner_delay,
-      selector = spinner_selector
-    ),
-    pulseOptions(
-      background = pulse_background,
-      height = pulse_height,
-      speed = pulse_speed
-    )
+  pulse_opts <- pulseOptions(
+    background = pulse_background,
+    height = pulse_height,
+    speed = pulse_speed
+  )
+
+  spinner_opts <- spinnerOptions(
+    type = spinner_type,
+    color = spinner_color,
+    size = spinner_size,
+    delay = spinner_delay,
+    selector = spinner_selector
   )
 
-  dropNulls(res)
+  if (!is.null(spinner_opts) && is.null(spinner_selector)) {
+    if (!is.null(pulse_opts)) {
+      abort(c(
+        "`spinner_selector` must be provided when customizing both the pulse and the spinner.",
+        "i" = "`spinner_selector = NULL` returns inline styles for spinner customizations."
+      ))
+    }
+    return(spinner_opts)
+  }
+
+  dropNulls(tagList(spinner_opts, pulse_opts))
 }
 
 
-# TODO: allow for customization of the spinner type.
-spinnerOptions <- function(color = NULL, size = NULL, delay = NULL, selector = NULL) {
-  if (is.null(color) && is.null(size) && is.null(delay) && is.null(selector)) {
+spinnerOptions <- function(
+  type = NULL,
+  color = NULL,
+  size = NULL,
+  delay = NULL,
+  selector = ":root"
+) {
+  if (is.null(type) && is.null(color) && is.null(size) && is.null(delay)) {
     return(NULL)
   }
 
+  if (!is.null(type)) {
+    if (!rlang::is_string(type)) {
+      spinners <- paste0(.spinner_types, collapse = '", "')
+      abort(
+        sprintf(
+          "`spinner_type` must be a string. Choose from \"%s\".",
+          spinners
+        )
+      )
+    }
+    if (grepl("^(http|[./])", type)) {
+      type <- sprintf("url('%s')", type)
+    } else {
+      type <- rlang::arg_match(type, .spinner_types)
+      type <- sprintf("url('spinners/%s.svg')", type)
+    }
+  }
+
   # Options controlled via CSS variables.
-  css_vars <- paste0(c(
-    if (!is.null(color)) sprintf("--shiny-spinner-color: %s", htmltools::parseCssColors(color)),
-    if (!is.null(size)) sprintf("--shiny-spinner-size: %s", htmltools::validateCssUnit(size)),
-    if (!is.null(delay)) sprintf("--shiny-spinner-delay: %s", delay)
-  ), collapse = ";")
+  css_vars <- htmltools::css(
+    `--shiny-spinner-url` = type,
+    `--shiny-spinner-color` = htmltools::parseCssColors(color),
+    `--shiny-spinner-size` = htmltools::validateCssUnit(size),
+    `--shiny-spinner-delay` = delay
+  )
```

Comment:
`htmltools::css()` is really good at this, but maybe there's a reason you were doing it more by hand?